src/chainstate/stacks/boot/exit-at-rc.clar

```diff
@@ -0,0 +1,199 @@
+        (list-indexes (list u0 u1 u2 u3 u4 u5 u6 u7 u8 u9 u10 u11))
+        ;; a list of the reward cycles the voter's vote is valid for, corresponding to the cycles their stx is locked for
+        (voting-reward-cycles (map get-voting-reward-cycles list-indexes lock-period-list first-reward-cycle-list ))
+    )
+        ;; Check that the amount stacked is positive
+        (asserts! (> amount-stacked u0) (err ERR_AMOUNT_NOT_POSITIVE))
+
+        ;; Check that the proposed exit rc is allowable
+        (asserts! (>= proposed-exit-rc (var-get absolute-minimum-exit-rc)) (err ERR_INVALID_PROPOSED_RC))
+        (asserts! (>= proposed-exit-rc (+ current-reward-cycle MINIMUM_RC_BUFFER_FROM_PRESENT)) (err ERR_INVALID_PROPOSED_RC))
+        (asserts! (<= proposed-exit-rc (+ current-reward-cycle MAXIMUM_RC_BUFFER_FROM_PRESENT)) (err ERR_INVALID_PROPOSED_RC))
+
+        ;; Check that the voter does not have an outstanding vote for this reward cycle
+        (match (map-get? voter-state {address: tx-sender})
+            voter-info (asserts! (>= current-reward-cycle (get expiration-reward-cycle voter-info))
+                (err ERR_PREVIOUS_VOTE_VALID))
+            ;; no existing state
+            true
+        )
+
+        ;; Check that the caller is allowed
+        (asserts! (is-eq tx-sender contract-caller) (err ERR_UNAUTHORIZED_CALLER))
+
+        ;; Modify the voter-state map
+        (map-set voter-state { address: tx-sender }
+            { proposed-rc: proposed-exit-rc, amount: amount-stacked, expiration-reward-cycle: stacking-expiration })
+
+        ;; Modify the rc-proposal-votes map - need to loop from curr rc to expiration rc
+        (map add-to-rc-proposal-map voting-reward-cycles proposed-rc-list amount-stacked-list)
+
+        (ok true)
+    )
+)
+
+;; This function is used by miners to veto a proposed exit reward cycle. The veto period is active the reward cycle
+;; after a vote is confirmed.
+;; Note: a miner can send in a veto in the block after the one they mined, and they can't include multiple of these
+;; transactions in a block.
+(define-public (veto-exit-rc (proposed-exit-rc uint))
+    (let (
+        (current-reward-cycle (unwrap-panic (current-pox-reward-cycle)))
+        (curr-vetoes (default-to u0 (get vetoes (map-get? rc-proposal-vetoes { proposed-rc: proposed-exit-rc, curr-rc: current-reward-cycle } ))))
+        (last-miner (unwrap! (get-block-info? miner-address (- block-height u1))
+                    (err ERR_FETCHING_BLOCK_INFO)))
+        (vetoed (default-to false (get vetoed (map-get? exercised-veto { proposed-rc: proposed-exit-rc, veto-height: block-height }))))
+    )
+
+    ;; a miner can only veto once per block
+    (asserts! (not vetoed) (err ERR_ALREADY_VETOED))
+
+    ;; a miner can only veto if they mined the previous block
+    (asserts! (is-eq contract-caller last-miner) (err ERR_UNAUTHORIZED_CALLER))
```

Comment:
This feels fragile.  It looks like if I'm a miner and I'm vetoing, then if the next miner is not vetoing, they could simply ignore my transaction and block my veto.  Is there a way to instead permit a miner to vote at any point in time prior to this reward cycle, and submit evidence that they were the miner for a given block (or blocks)?